### PR TITLE
3.6 - 14359 - CountStack() added to beanshell for piece expressions

### PR DESCRIPTION
**Pull request by Cattlesquat**
CountStack("Prop") returns the number of pieces in the same stack which have any non-blank value for the property "Prop".

CountStack("") with a blank property simply returns the number of pieces in the same stack.

### CODE REVIEW

File: vassal-app/src/main/resources/VASSAL/script/init_expression.bsh [hunk 1]
```diff
@@ -86,6 +86,18 @@ SumStack(String property) {
...
+ * stack that *contain* the target property
+ *
+ * Example: CountStack("Strength")
+ */
+CountStack(String property) {
+  _interp = this.caller.namespace.getVariable("_interp");
+  _source = this.caller.namespace.getVariable("_source");
+  return unwrap(_interp.countStack(property, _source));
+}
+
```
Comment (BrentEaston) [inline vassal-app/src/main/resources/VASSAL/script/init_expression.bsh:100]: Create a second version

CountStack() {
    return CountStack("");
}

File: vassal-app/src/main/java/VASSAL/configure/BeanShellFunctionMenu.java [hunk 2]
```diff
@@ -152,6 +152,7 @@ public BeanShellFunctionMenu(EditablePiece target, BeanShellExpressionConfigurer
 
     final JMenu countMenu = new JMenu(Resources.getString("Editor.BeanShell.sumcount"));
     addFunction(countMenu, "SumStack", Resources.getString("Editor.BeanShell.sum1"), new String[] { Resources.getString("Editor.BeanShell.property_name") }, "(name)"); //NON-NLS
+    addFunction(countMenu, "CountStack", Resources.getString("Editor.BeanShell.sum6"), new String[] { Resources.getString("Editor.BeanShell.property_name") }, "(name)"); //NON-NLS
```
Comment (BrentEaston) [inline vassal-app/src/main/java/VASSAL/configure/BeanShellFunctionMenu.java:155]: Add an additional entry for CountStack() with no parameters

File: vassal-app/src/main/resources/VASSAL/i18n/Editor.properties [hunk 3]
```diff
@@ -260,6 +260,7 @@ Editor.BeanShell.sum2=Sum the values of the named property in matching pieces on
 Editor.BeanShell.sum3=Sum the values of the named property in matching pieces on the named map
 Editor.BeanShell.sum4=Count the number of matching pieces on all maps
 Editor.BeanShell.sum5=Count the number of matching pieces on the named map
+Editor.BeanShell.sum6=Count the number of pieces with any non-blank value of named property, in the same stack. Or if blank property passed, returns the total number of pieces in the stack.
```
Comment (BrentEaston) [inline vassal-app/src/main/resources/VASSAL/i18n/Editor.properties:263]: Extra description for the no-args version. Remove the 'Blank property' blah for the 1-arg version. The code won't change, but we will direct designers to use the no-arg version if they just want a straight stack count.

File: vassal-doc/src/main/readme-referencemanual/ReferenceManual/Expression.adoc [hunk 4]
```diff
@@ -165,6 +166,8 @@ Count(expr, map) + |{SumStack("Attack")} +
 {Count("{ Nation == \"$Nation$\" }")} +
 {Count("{ CombatValue > 0 }", "BattleMap") } + |*SumStack(prop)* returns the total of the specified property in all pieces in the same stack as this one.
 
+*CountStack(prop)* returns the number of pieces in the same stack as this one, that have a non-blank value for the named property (or a count of the total pieces in the stack if a blank property is sent)
```
Comment (BrentEaston) [inline vassal-doc/src/main/readme-referencemanual/ReferenceManual/Expression.adoc:169]: No-arg and 1-arg versions

**Review by BrentEaston — CHANGES_REQUESTED** [review]
Looks good, just split it into a 1-arg version for checking a property and a no-arg version for a straight stack count. The guts of you code won't change. CountStack() will do the same as CountStack("").

**Comment by Cattlesquat** [comment]
@BrentEaston Good stuff - I believe I have done "all the things"!

**Pull request merged**